server/src/main/java/org/opensearch/cluster/service/Throttler.java

```diff
@@ -0,0 +1,117 @@
+
+    private volatile boolean throttlingEnabled;
+
+    public Throttler(final boolean throttlingEnabled) {
+        this.throttlingEnabled = throttlingEnabled;
+    }
+
+    /**
+     * Method to acquire permits for a key type.
+     * If throttling is disabled, it will always return True,
+     * else it will return true if permits can be acquired within threshold limits.
+     *
+     * If threshold is not set for key then also it will return True.
+     *
+     * @param type Key for which we want to acquire permits.
+     * @param permits Number of permits to acquire.
+     * @return boolean representing was it able to acquire the permits or not.
+     */
+    public boolean acquire(final T type, final int permits) {
+        assert permits > 0;
+        AdjustableSemaphore semaphore = semaphores.get(type);
+        if (throttlingEnabled && Objects.nonNull(semaphore)) {
+            return semaphore.tryAcquire(permits);
+        }
+        return true;
+    }
+
+    /**
+     * Release the given permits for given type.
+     *
+     * @param type key for which we want to release permits.
+     * @param permits number of permits to release.
+     */
+    public void release(final T type, final int permits) {
+        assert permits > 0;
+        AdjustableSemaphore semaphore = semaphores.get(type);
+        if (throttlingEnabled && Objects.nonNull(semaphore)) {
+            semaphore.release(permits);
+            assert semaphore.availablePermits() <= semaphore.getMaxPermits();
+        }
+    }
+
+    /**
+     * Update the Threshold for throttling for given type.
+     *
+     * @param key Key for which we want to update limit.
+     * @param newLimit Updated limit.
+     */
+    public synchronized void updateThrottlingLimit(final T key, final Integer newLimit) {
+        assert newLimit >= 0;
+        if (semaphores.containsKey(key)) {
+            semaphores.get(key).setMaxPermits(newLimit);
+        } else {
+            semaphores.put(key, new AdjustableSemaphore(newLimit, true));
+        }
+    }
+
+    /**
+     * Remove the threshold for given key.
+     * Throttler will no longer do throttling for given key.
+     *
+     * @param key Key for which we want to remove throttling.
+     */
+    public synchronized void removeThrottlingLimit(final T key) {
+        assert semaphores.containsKey(key);
+        semaphores.remove(key);
+    }
+
+    /**
+     * Set flag for enabling/disabling the throttling logic.
+     * Clear the state of previous semaphores with each update.
+     *
+     * @param throttlingEnabled flag repressing enabled/disabled throttling.
+     */
+    public synchronized void setThrottlingEnabled(final boolean throttlingEnabled) {
+        this.throttlingEnabled = throttlingEnabled;
+        for (T key : semaphores.keySet()) {
+            semaphores.put(key, new AdjustableSemaphore(semaphores.get(key).getMaxPermits(), true));
+        }
```

Comment:
In case throttlingEnabled is false, you dont need to do this.

Reply:
This method will be removed, as we will remove the overall throttling flag.